presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/GroupByWithConstants.java

```diff
@@ -0,0 +1,148 @@
+    {
+        if (!isOptimizeConstantGroupingKeys(context.getSession())) {
+            return Result.empty();
+        }
+
+        // for each variable references in grouping keys, check if the CHILD expression defines them as constants
+        ProjectNode child = captures.get(CHILD);
+        List<VariableReferenceExpression> outputVariables = parent.getOutputVariables();
+
+        Map<VariableReferenceExpression, RowExpression> constChildVars = extractConstVars(child, outputVariables);
+        Map<VariableReferenceExpression, RowExpression> constGroupingKeys = new HashMap<>();
+
+        List<VariableReferenceExpression> groupingKeys = parent.getGroupingKeys();
+        for (VariableReferenceExpression key : groupingKeys) {
+            if (constChildVars.containsKey(key)) {
+                constGroupingKeys.put(key, constChildVars.get(key));
+            }
+        }
+
+        if (constGroupingKeys.isEmpty()) {
+            return Result.empty();
+        }
+
+        List<VariableReferenceExpression> newGroupingKeys =
+                groupingKeys.stream()
+                        .filter(key -> !constChildVars.containsKey(key))
+                        .collect(Collectors.toList());
+
+        AggregationNode newAgg = new AggregationNode(
+                parent.getSourceLocation(),
+                parent.getId(),
+                child,
+                parent.getAggregations(),
+                AggregationNode.singleGroupingSet(newGroupingKeys),
+                ImmutableList.of(),
+                parent.getStep(),
+                parent.getHashVariable(),
+                parent.getGroupIdVariable());
+
+        Map<VariableReferenceExpression, RowExpression> remainingVars =
+                outputVariables.stream()
+                        .filter(var -> !constChildVars.containsKey(var))
+                        .collect(Collectors.toMap(Function.identity(), Function.identity()));
+
+        Assignments.Builder assignments = Assignments.builder();
+        assignments.putAll(constChildVars);
+        assignments.putAll(remainingVars);
+        return Result.ofPlanNode(
+                new ProjectNode(
+                parent.getSourceLocation(),
+                context.getIdAllocator().getNextId(),
+                newAgg,
+                assignments.build(),
+                child.getLocality()));
+    }
+
+    private static Map<VariableReferenceExpression, RowExpression> extractConstVars(ProjectNode child, List<VariableReferenceExpression> outputVariables)
+    {
+        Map<VariableReferenceExpression, RowExpression> result = new HashMap<>();
+
+        Set<Map.Entry<VariableReferenceExpression, RowExpression>> childAssighnments = child.getAssignments().entrySet();
+        for (Map.Entry<VariableReferenceExpression, RowExpression> entry : childAssighnments) {
+            if (isConstant(entry.getValue()) && outputVariables.contains(entry.getKey())) {
```

Comment:
I believe we have a isConst method in ExpressionUtils?

Reply:
yes, thank you!

Reply:
done: the utility function expects Expression, I added a wrapper below